### PR TITLE
Implement Disposable hierarchy for most of the listeners

### PR DESCRIPTION
**Pull request by TimvdLippe**
Build on top of #276, only new commit is https://github.com/TestRoots/watchdog/commit/4c414364dbd7b2afd506c5e6fdd6706ded36294c

This makes sure most of the listeners related to editor events are properly cleaning up and using the official `Disposable` interface. By implementing `Disposable`, we can use `Disposer.register` which means the `Disposable` hierarchy cleans up automatically once we call `Disposer.dispose`.

This PR also includes some auto-fixes as reported by IntelliJ (`F2` and then `alt + Enter`).

### CODE REVIEW

File: intellij/src/nl/tudelft/watchdog/intellij/logic/InitializationManager.java [hunk 1]
```diff
@@ -51,7 +53,13 @@
...
         // Initialize persisters
         // Double getPath() because they are different methods on different objects
-        File baseFolder = new File(PluginManager.getPlugin(PluginId.findId("nl.tudelft.watchdog")).getPath().getPath());
+        final IdeaPluginDescriptor plugin = PluginManager.getPlugin(PluginId.findId("nl.tudelft.watchdog"));
+
+        if (plugin == null) {
+            throw new IllegalArgumentException("Plugin id \"nl.tudelft.watchdog\" could not be found in the list of installed plugins.");
+        }
+
+        File baseFolder = new File(plugin.getPath().getPath());
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/InitializationManager.java]: see comment above on double getpath getpath

File: intellij/src/nl/tudelft/watchdog/intellij/logic/InitializationManager.java [hunk 2]
```diff
@@ -51,7 +53,13 @@
     private InitializationManager(Project project) {
         // Initialize persisters
         // Double getPath() because they are different methods on different objects
-        File baseFolder = new File(PluginManager.getPlugin(PluginId.findId("nl.tudelft.watchdog")).getPath().getPath());
+        final IdeaPluginDescriptor plugin = PluginManager.getPlugin(PluginId.findId("nl.tudelft.watchdog"));
+
+        if (plugin == null) {
+            throw new IllegalArgumentException("Plugin id \"nl.tudelft.watchdog\" could not be found in the list of installed plugins.");
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/InitializationManager.java]: can we get the pluginid programmatically?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java [hunk 3]
```diff
@@ -1,53 +1,70 @@
...
 import java.util.HashMap;
 import java.util.Map;
 
-public class EditorWindowListener implements EditorFactoryListener {
+public class EditorWindowListener implements EditorFactoryListener, Disposable {
 
     private EditorFocusListener focusListener;
 
-    private Map<Editor, EditorListener> editorListenerMap = new HashMap<Editor, EditorListener>();
+    private Map<Editor, Disposable> editorListenerMap = new HashMap<>();
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java]: Can you adapt the name of the hashmap and explain the tree concept a bit?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/InitializationManager.java (hunk 2 above)
Comment (TimvdLippe) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/InitializationManager.java]: This is specified in the `META-INF/plugin.xml` which is quite tedious to parse as XML. If we want to improve this, I would put it in a separate PR.

**Comment by Inventitech** [comment]
I'll have a look at this once #276 is merged.

**Comment by Inventitech** [comment]
Can you rebase this PR against the current master? It's a bit hard to see the changes of this PR currently.

**Comment by TimvdLippe** [comment]
Done

File: intellij/src/nl/tudelft/watchdog/intellij/logic/InitializationManager.java [hunk 4]
```diff
@@ -49,10 +51,16 @@
      * Private constructor.
      */
     private InitializationManager(Project project) {
-        // Initialize persisters
+        final IdeaPluginDescriptor plugin = PluginManager.getPlugin(PluginId.findId("nl.tudelft.watchdog"));
+
+        if (plugin == null) {
+            throw new IllegalArgumentException("Plugin id \"nl.tudelft.watchdog\" could not be found in the list of installed plugins.");
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/InitializationManager.java]: Use magic string from above?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/InitializationManager.java [hunk 5]
```diff
@@ -49,10 +51,16 @@
      * Private constructor.
      */
     private InitializationManager(Project project) {
-        // Initialize persisters
+        final IdeaPluginDescriptor plugin = PluginManager.getPlugin(PluginId.findId("nl.tudelft.watchdog"));
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/InitializationManager.java]: Extract magic string?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/event/listeners/BreakpointListener.java [hunk 6]
```diff
@@ -24,7 +24,7 @@
  * {@link BreakpointEventBase} is generated and given to the
  * {@link TrackingEventManager}.
  */
-public class BreakpointListener implements XBreakpointListener {
+public class BreakpointListener implements XBreakpointListener<XBreakpoint<?>> {
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/event/listeners/BreakpointListener.java]: Do we really need the additional `<XBreakpoint<?>>`?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java [hunk 7]
```diff
@@ -1,53 +1,76 @@
...
 import java.util.Map;
 
-public class EditorWindowListener implements EditorFactoryListener {
+public class EditorWindowListener implements EditorFactoryListener, Disposable {
 
     private EditorFocusListener focusListener;
 
-    private Map<Editor, EditorListener> editorListenerMap = new HashMap<Editor, EditorListener>();
+    /**
+     * Every editor that is created is associated with a tree of disposables.
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java]: Try to avoid passive here. "We associate every created editor ..."

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java [hunk 8]
```diff
@@ -1,53 +1,76 @@
...
+     * The parent Disposable is registered with each associated listener to
+     * events of the editor. By using Disposable for this structure, we can
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java]: Link to Disposable?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java [hunk 9]
```diff
@@ -1,53 +1,76 @@
...
+        markupModel.addMarkupModelListener(parentDisposable, new IntelliJMarkupModelListener(trackingEventManager));
+
+        editorDisposableTreeMap.put(editor, parentDisposable);
     }
 
     @Override
-    public void editorReleased(EditorFactoryEvent editorFactoryEvent) {
-        if(!editorBelongsToThisProject(editorFactoryEvent)) {
+    public void editorReleased(@NotNull EditorFactoryEvent editorFactoryEvent) {
+        if (editorDoesNotBelongToProject(editorFactoryEvent)) {
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java]: Why can there not be an NPE anymore?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java [hunk 10]
```diff
@@ -1,53 +1,76 @@
...
-            return false;
-        }
+    private boolean editorDoesNotBelongToProject(EditorFactoryEvent editorFactoryEvent) {
+        return !project.equals(editorFactoryEvent.getEditor().getProject());
+    }
+
+    @Override
+    public void dispose() {
+        editorDisposableTreeMap.keySet()
+                .forEach(editor ->Disposer.dispose(editorDisposableTreeMap.remove(editor)));
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java]: Nice. `editor -> Disposer`?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/GeneralActivityListener.java [hunk 11]
```diff
@@ -10,7 +11,7 @@
  * A listener that determines whether there was general activity in the IntelliJ
  * window.
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/GeneralActivityListener.java]: Mention self-desctruction capabilities via dispose()?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java [hunk 12]
```diff
@@ -20,67 +21,67 @@
  * Sets up the listeners for IntelliJ UI events and registers the shutdown
  * listeners.
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java]: Mention self-destruction capabilities via dispose()?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java [hunk 13]
```diff
@@ -20,67 +21,67 @@
...
      */
-    public void attachListeners() {
+    private void attachListeners() {
         WatchDogEventType.START_IDE.process(this);
 
         connection.subscribe(ApplicationActivationListener.TOPIC,
                 new IntelliJActivationListener());
-        activityListener = new GeneralActivityListener(project.getName());
-        EditorFactory.getInstance().addEditorFactoryListener(editorWindowListener, parent);
+        Disposer.register(this, new GeneralActivityListener(project.getName()));
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java]: OK, I don't understand how this line works :-). Can you explain it to me? Thanks. Why do you have to register this listener, but not the others?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJMarkupModelListener.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
+package nl.tudelft.watchdog.intellij.logic.ui.listeners;
+
+import com.intellij.openapi.editor.ex.RangeHighlighterEx;
+import com.intellij.openapi.editor.impl.event.MarkupModelListener;
+import nl.tudelft.watchdog.core.logic.event.TrackingEventManager;
+import org.jetbrains.annotations.NotNull;
+
+public class IntelliJMarkupModelListener implements MarkupModelListener {
```
Comment (Inventitech) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJMarkupModelListener.java]: This should be part of another PR, right?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/event/listeners/BreakpointListener.java (hunk 6 above)
Comment (TimvdLippe) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/event/listeners/BreakpointListener.java]: Yes, else it complains about missing the generics specification. E.g. `List` vs `List<Type>`

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java (hunk 8 above)
Comment (TimvdLippe) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java]: It is already in the type information, but added it nonetheless.

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java (hunk 9 above)
Comment (TimvdLippe) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java]: Because we now compare projects, rather than names. This means that if one of the arguments is `null`, it will compare that to an actual object.

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/GeneralActivityListener.java (hunk 11 above)
Comment (TimvdLippe) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/GeneralActivityListener.java]: That is already deducible from the `implements Disposable`, so I would say this is not necessary. WDYT?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java (hunk 13 above)
Comment (TimvdLippe) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java]: Most of the listener APIs in IntelliJ accept a `Disposable` argument. Whenever you want to add a Listener to something, you have to supply a `Disposable`. Whenever this `Disposable` is disposed, the listener is automatically removed.

Sadly, not all listener methods follow this pattern. `addEditorFactoryListener` does, but `addAWTEventListener` as used in `GeneralActivityListener` does not.

Lastly, the messageBus (`connection`) is already registered in the constructor with `connection = messageBus.connect(); Disposer.register(this, connection);`

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJMarkupModelListener.java (hunk 14 above)
Comment (TimvdLippe) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJMarkupModelListener.java]: I included it to show how a new class would fit in this model. It will be changed in the next PR, but I can remove it if you want to?

Comment (Inventitech) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJMarkupModelListener.java]: Yeah, I think that would be cleaner. Thanks

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/GeneralActivityListener.java (hunk 11 above)
Comment (Inventitech) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/GeneralActivityListener.java]: I'd prefer to have this explicit.

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java (hunk 9 above)
Comment (Inventitech) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java]: And project can never be null? (This comment should really be displayed a couple lines below)

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java (hunk 13 above)
Comment (Inventitech) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java]: Wow, this non-uniformity needs to be documented somewhere

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java (hunk 9 above)
Comment (TimvdLippe) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java]: Per the API docs, our own `project` can not be null, but the project in `editorFactoryEvent` can be. Therefore, invoking `equals` on `project` is safe, but if we would flip it, it would not be safe.

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java (hunk 13 above)
Comment (TimvdLippe) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java]: Yup, sadly the documentation of IntelliJ itself is lackluster. I am not aware of a style guide that they use for these kind of API designs. Might be a legacy issue as well.

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java (hunk 9 above)
Comment (Inventitech) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java]: OK

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java (hunk 13 above)
Comment (Inventitech) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java]: Can you integrate parts of your comment above as a comment/documentation in the source code?

File: intellij/src/nl/tudelft/watchdog/intellij/logic/event/listeners/BreakpointListener.java (hunk 6 above)
Comment (Inventitech) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/event/listeners/BreakpointListener.java]: Ok

**Comment by Inventitech** [comment]
Only one TODO left to acceptance :)

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java (hunk 13 above)
Comment (TimvdLippe) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/IntelliJListener.java]: Done :+1:

**Comment by TimvdLippe** [comment]
@Inventitech done

**Comment by Inventitech** [comment]
Merge!
